File: packages/sdk-python/examples/create_and_send_invoice.py
"""
Create a customer, bill them from the catalog, finalize and email the invoice.

    INVOICE_AI_API_KEY=inv_live_… uv run python examples/create_and_send_invoice.py

Point INVOICE_AI_BASE_URL at http://localhost:3000/api/v1 to try it locally.
"""

from __future__ import annotations

from invoice_ai import InvalidStateError, InvoiceAI, ValidationError, format_money, to_minor


def main() -> None:
    client = InvoiceAI()  # reads INVOICE_AI_API_KEY

    customer = client.customers.create(name="Acme Industries", email="delivered@resend.dev")
    product = client.products.create(name="Consulting retainer")
    price = client.prices.create(product=product.id, currency="USD", unit_amount=to_minor("2500.00", "USD"))

    try:
        draft = client.invoices.create(
            customer=customer.id,
            items=[
                {"price": price.id, "quantity": 1},
                {"description": "Onboarding workshop", "unit_amount": to_minor("500", "USD")},
            ],
            days_until_due=30,
        )
    except ValidationError as e:
        for field in e.fields:
            print(f"  {field['path']}: {field['message']}")
        raise

    invoice = client.invoices.finalize(draft.id)
    print(f"Finalized {invoice.number}: {format_money(invoice.total, invoice.currency)}")

    try:
        sent = client.invoices.send(invoice.id)
        print(f"Emailed to {sent.emailed_to}")
    except InvalidStateError as e:
        print(f"Could not send ({e.code}): {e.detail} — request {e.request_id}")

    # Every open invoice, across all pages.
    for inv in client.invoices.list(status="open"):
        print(inv.number, format_money(inv.amount_due, inv.currency))


if __name__ == "__main__":
    main()

File: packages/sdk-ts/examples/create-and-send-invoice.ts
/**
 * Create a customer, invoice them, finalize and email the invoice.
 *
 *   INVOICE_AI_API_KEY=inv_live_… npx tsx examples/create-and-send-invoice.ts
 *
 * Point INVOICE_AI_BASE_URL at a Prism mock or a local server to try it safely:
 *   npx @stoplight/prism-cli mock ../../api-docs/openapi.json
 *   INVOICE_AI_API_KEY=inv_live_test_x INVOICE_AI_BASE_URL=http://127.0.0.1:4010 npx tsx examples/create-and-send-invoice.ts
 */
import InvoiceAI, { InvalidStateError, ValidationError, formatMoney, toMinor } from '@horizonpay/invoice-ai'

const invoiceai = new InvoiceAI() // reads INVOICE_AI_API_KEY

async function main() {
  const customer = await invoiceai.customers.create({ name: 'Acme Industries', email: 'delivered@resend.dev' })

  const draft = await invoiceai.invoices.create({
    customer: customer.id,
    currency: 'USD',
    items: [{ description: 'Consulting retainer — September', quantity: 1, unit_amount: toMinor('2500.00', 'USD') }],
  })

  const invoice = await invoiceai.invoices.finalize(draft.id)
  await invoiceai.invoices.send(invoice.id)
  console.log(`Sent ${invoice.number} for ${formatMoney(invoice.total, invoice.currency)} to ${customer.email}`)

  // Every page of open invoices, fetched lazily.
  for await (const open of invoiceai.invoices.list({ status: 'open' })) {
    console.log(open.number, formatMoney(open.amount_due, open.currency))
  }
}

main().catch((err: unknown) => {
  if (err instanceof ValidationError) console.error('Fix these fields:', err.fields)
  else if (err instanceof InvalidStateError) console.error('Wrong state:', err.detail)
  else console.error(err)
  process.exitCode = 1
})
